### PR TITLE
Add worker to clean up stale DisruptionTarget condition

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is a requirement from #110959.

It is possible that controllers that add the DisruptionTarget condition and fail before being able to Delete a pod. When the controller restarts, it might take a different decision, leaving a stale condition. This worker, added to the disruption controller, clears this condition if a DeletionTimestamp is not added to the pod after 2 minutes.

#### Which issue(s) this PR fixes:

Ref kubernetes/enhancements#3329

#### Special notes for your reviewer:

The first commit introduces a clock interface to the controller in order to write the unit tests more accurately. I proposed it initially as a separate PR in #111447, but it seems small enough to be in this PR as a separate commit.

This functionality is purposely not gated. This is because if the feature gate PodDisruptionConditions is disabled, we still want to clear the condition.

#### Does this PR introduce a user-facing change?

```release-note
If a Pod has a DisruptionTarget condition with status=True for more than 2 minutes without getting a DeletionTimestamp, the control plane resets it to status=False
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/sig apps
/assign @liggitt @soltysh 

cc @mimowo

**Review by soltysh — APPROVED** [review]
minor nits, controller changes 
/lgtm
/approve

**Comment by alculquicondor** [comment]
/hold
to prevent premature merge

**Comment by liggitt** [comment]
Unfortunately, I don't have bandwidth to review the impl here, will defer to apps reviewers. The main thing I'd be looking for here is testing and isolation of the new code, and if it's connected to a new feature, what the fallback/disable plan is if we discover the new worker has issues after release.

File: pkg/controller/disruption/disruption.go
Comment (alculquicondor) [reply pkg/controller/disruption/disruption.go:656]: I somewhat disagree on this. This is already an `error`, so it shouldn't need the word error again. And if some day the implementation of HandleError changes and wraps the given error, it wouldn't read well.

But sure, given the current implementation of HandleError, your recommendation makes sense.

File: pkg/controller/disruption/disruption.go
Comment (alculquicondor) [reply pkg/controller/disruption/disruption.go:566]: Oops, left over from my debugging.

**Review by mimowo — COMMENTED** [review]
This PR introduces a user-facing behavior on its own - and it is not gated, so it seems it should have a dedicated release note.

File: pkg/controller/disruption/disruption.go
Comment (alculquicondor) [reply pkg/controller/disruption/disruption.go:760]: Used the word `Reset` instead.

File: test/integration/disruption/disruption_test.go
Comment (alculquicondor) [reply test/integration/disruption/disruption_test.go:689]: Done

File: test/integration/disruption/disruption_test.go
Comment (alculquicondor) [reply test/integration/disruption/disruption_test.go:729]: I'm like 99% sure that it would be there, but I added it just to be safe.

File: test/integration/disruption/disruption_test.go
Comment (alculquicondor) [reply test/integration/disruption/disruption_test.go:733]: It kind of ensures that the the controller or test code did a Delete when it shouldn't

File: cmd/kube-controller-manager/app/policy.go
Comment (alculquicondor) [reply cmd/kube-controller-manager/app/policy.go:53]: In general, I would avoid adding a function that is only used for tests, like `NewDisruptionControllerInternal`, but since we have to add it, I prefer that the non internal function has as much parameters as possible, so usage is more explicit.

Comment (mimowo) [reply cmd/kube-controller-manager/app/policy.go:53]: I see, but isn't clock by nature a technical detail for testing rather than part of the public API one needs to call to construct an object? Also, another consideration is that the PR requires approval from cmd/OWNERS which can be easily avoided.

File: test/integration/disruption/disruption_test.go
Comment (mimowo) [reply test/integration/disruption/disruption_test.go:729]: I think it would be easier to read if the waiting for DeletionTimestamp !=nil is just after deleting.

File: test/integration/disruption/disruption_test.go
Comment (mimowo) [reply test/integration/disruption/disruption_test.go:733]: I assume you mean didn't do a Delete when it shouldn't. You may leave it, but it feels unnecessary to me, we could potentially be checking many other things to check that the controller or test code didn't do.

**Comment by mimowo** [comment]
/lgtm 
left with remarks which aren't blocking. I will revoke the lgtm if updated.

File: test/integration/disruption/disruption_test.go
Comment (alculquicondor) [reply test/integration/disruption/disruption_test.go:732]: Sure

**Review by janetkuo — APPROVED** [review]
/lgtm

File: pkg/controller/disruption/disruption.go
Comment (aojea) [inline pkg/controller/disruption/disruption.go:725]:
```
	startTime := dc.clock.Now()
```

File: pkg/controller/disruption/disruption.go
Comment (aojea) [inline pkg/controller/disruption/disruption.go:731]:
```
dc.clock.Since(startTime)
```

File: pkg/controller/disruption/disruption.go
Comment (aojea) [inline pkg/controller/disruption/disruption.go:739]: it seems you don't have a limit of retries, is that expected?

File: pkg/controller/disruption/disruption_test.go
Comment (aojea) [inline pkg/controller/disruption/disruption_test.go:181]: from where `ctx.Done()` is coming after removing this?

File: pkg/controller/disruption/disruption.go
Comment (alculquicondor) [reply pkg/controller/disruption/disruption.go:739]: Correct. This is the same we do for any controller. We just stop if the pod is not found.

File: pkg/controller/disruption/disruption_test.go
Comment (alculquicondor) [reply pkg/controller/disruption/disruption_test.go:181]: It's coming from the test. See `newFakeDisruptionControllerWithTime`

**Comment by alculquicondor** [comment]
/assign @lavalamp
for `staging/src/k8s.io/client-go/util/workqueue/rate_limiting_queue.go`

**Review by soltysh — APPROVED** [review]
/lgtm

**Pull request merged**
